### PR TITLE
CDAP-2039 Generating RunId in the AbstractProgramRuntimeService and pass...

### PR DESCRIPTION
**Pull request by sagarkapare**
...ing it as ProgramOptions to the programs.

Jira: https://issues.cask.co/browse/CDAP-2039

Move the RunId generation logic to the AbstractProgramRuntimeService. All programs that are started explicitly will now receive the RunId as a part of ProgramOptions. 
When the program is started as a part of the Workflow, workflow will create the new RunId and then pass it as a part of ProgramOptions to the underlying program.

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT1459-2

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT1459-3

**Comment by sagarkapare** [comment]
@chtyim Please take a look. Thanks!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java [hunk 1]
```diff
@@ -922,6 +922,38 @@ public Void apply(AppMds mds) throws Exception {
     });
   }
 
+  @Override
+  public Map<String, String> getRunIdMappings() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java]: Should return `RunIdMapping`?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java [hunk 2]
```diff
@@ -536,4 +540,36 @@ private MDSKey getNamespaceKey(@Nullable String name) {
     }
     return builder.build();
   }
+
+  private MDSKey getRunIdMappingKey() {
+    return new MDSKey.Builder()
+      .add(TYPE_RUNID_MAPPING)
+      .build();
+  }
+  public Map<String, String> getRunIdMappings() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java]: Empty line missing. Javadoc is needed for public methods.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java [hunk 3]
```diff
@@ -54,12 +57,26 @@ protected AbstractProgramRuntimeService(ProgramRunnerFactory programRunnerFactor
...
+    ProgramOptions optionsWithRunId = createProgramOptionsWithRunId(options);
+    final SimpleRuntimeInfo runtimeInfo = new SimpleRuntimeInfo(runner.run(program, optionsWithRunId), program);
     addRemover(runtimeInfo);
     runtimeInfos.put(runtimeInfo.getType(), runtimeInfo.getController().getRunId(), runtimeInfo);
     return runtimeInfo;
   }
 
+  private ProgramOptions createProgramOptionsWithRunId(ProgramOptions options) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    for (Map.Entry<String, String> systemArgument : options.getArguments().asMap().entrySet()) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java]: Use builder.putAll().

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java [hunk 4]
```diff
@@ -127,6 +134,40 @@
...
 
+  private synchronized Map<String, String> getRunIdMap() {
+    if (runIdMap == null) {
+      runIdMap = store.getRunIdMappings();
+    }
+    return runIdMap;
+  }
+
+  @Override
+  protected synchronized RuntimeInfo addRemover(final RuntimeInfo runtimeInfo) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java]: Rename the method to `programStarted`.

**Comment by sagarkapare** [comment]
Discussed this with @chtyim and more refactoring required for storing the mapping between Twill RunId and CDAP RunId. Instead of creating new table, we decided to store this mapping as a part of the RunRecord itself. So, in distributed mode RunRecord for the programs would contain 2 ids - one is generated by the CDAP and another is generated by Twill. Working on making these changes.

**Comment by sagarkapare** [comment]
Updated the PR. Build is running here - http://builds.cask.co/browse/CDAP-DUT1459

**Comment by sagarkapare** [comment]
@chtyim Please take a look again. Thanks!

**Comment by sagarkapare** [comment]
Not able to update this PR. Closing it for now and open a new one.

**Pull request closed without merging**